interfaces/policy/policy.go

```diff
@@ -30,6 +30,50 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+// InstallCandidateMinimalCheck represents a candidate snap installed with --dangerous flag that should pass minimum checks
+// against snap type (if present). It doesn't check interface attributes.
+type InstallCandidateMinimalCheck struct {
+	Snap            *snap.Info
+	BaseDeclaration *asserts.BaseDeclaration
+	Model           *asserts.Model
+	Store           *asserts.Store
+}
+
+func (ic *InstallCandidateMinimalCheck) checkSlotRule(slot *snap.SlotInfo, rule *asserts.SlotRule) error {
+	if hasConstraints, err := checkMinimalSlotInstallationConstraints(ic, slot, rule.DenyInstallation); hasConstraints && err == nil {
+		return fmt.Errorf("installation denied by %q slot rule of interface %q", slot.Name, slot.Interface)
+	}
+
+	if hasConstraints, err := checkMinimalSlotInstallationConstraints(ic, slot, rule.AllowInstallation); hasConstraints && err != nil {
```

Comment:
here afaict if err != nil then hasConstraints is true. For me it would be clearer to omit checking hasConstraints then.

Reply:
In the past  if there were no snap type constraints at all,  implicitly there was an expectation that the snap is an "app", I'm not sure that is true anymore, we added a couple of such interfaces that can make sense for a gadget too. Maybe add a `// TODO check that the snap is an app or gadget if allow-installation had no slot-snap-type constraints`